subsys/bluetooth/audio/mics_client.c

```diff
@@ -0,0 +1,536 @@
+				if (mics_client_cb != NULL &&
+				    mics_client_cb->discover != NULL) {
+					mics_client_cb->discover(conn, err, 0);
+				}
+			}
+		} else {
+			if (mics_client_cb != NULL &&
+			    mics_client_cb->discover != NULL) {
+				mics_client_cb->discover(conn, err, 0);
+			}
+		}
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("[ATTRIBUTE] handle 0x%04X", attr->handle);
+
+	if (params->type == BT_GATT_DISCOVER_CHARACTERISTIC) {
+		struct bt_gatt_chrc *chrc = (struct bt_gatt_chrc *)attr->user_data;
+		struct bt_gatt_subscribe_params *sub_params = NULL;
+
+		if (bt_uuid_cmp(chrc->uuid, BT_UUID_MICS_MUTE) == 0) {
+			BT_DBG("Mute");
+			mics_inst->mute_handle = chrc->value_handle;
+			sub_params = &mics_inst->mute_sub_params;
+			sub_params->disc_params = &mics_inst->mute_sub_disc_params;
+		}
+
+		if (sub_params != NULL) {
+			int err;
+
+			/* With ccc_handle == 0 it will use auto discovery */
+			sub_params->ccc_handle = 0;
+			sub_params->end_handle = mics_inst->end_handle;
+			sub_params->value = BT_GATT_CCC_NOTIFY;
+			sub_params->value_handle = chrc->value_handle;
+			sub_params->notify = mute_notify_handler;
+
+			err = bt_gatt_subscribe(conn, sub_params);
+			if (err == 0) {
+				BT_DBG("Subscribed to handle 0x%04X",
+				       attr->handle);
+			} else {
+				BT_DBG("Could not subscribe to handle 0x%04X: %d",
+				       attr->handle, err);
+			}
+		}
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t primary_discover_func(struct bt_conn *conn,
+				     const struct bt_gatt_attr *attr,
+				     struct bt_gatt_discover_params *params)
+{
+	struct mics_instance *mics_inst = &mics_insts[bt_conn_index(conn)];
+
+	if (attr == NULL) {
+		BT_DBG("Could not find a MICS instance on the server");
+		if (mics_client_cb != NULL &&
+		    mics_client_cb->discover != NULL) {
+			mics_client_cb->discover(conn, -ENODATA, 0);
+		}
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("[ATTRIBUTE] handle 0x%04X", attr->handle);
+
+	if (params->type == BT_GATT_DISCOVER_PRIMARY) {
+		struct bt_gatt_service_val *prim_service =
+			(struct bt_gatt_service_val *)attr->user_data;
+		int err;
+
+		BT_DBG("Primary discover complete");
+		mics_inst->start_handle = attr->handle + 1;
+		mics_inst->end_handle = prim_service->end_handle;
+
+		/* Discover characteristics */
+		mics_inst->discover_params.uuid = NULL;
+		mics_inst->discover_params.start_handle = mics_inst->start_handle;
+		mics_inst->discover_params.end_handle = mics_inst->end_handle;
+		mics_inst->discover_params.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+		mics_inst->discover_params.func = mics_discover_func;
+
+		err = bt_gatt_discover(conn, &mics_inst->discover_params);
+		if (err != 0) {
+			BT_DBG("Discover failed (err %d)", err);
+			if (mics_client_cb != NULL &&
+			    mics_client_cb->discover != NULL) {
+				mics_client_cb->discover(conn, err, 0);
+			}
+		}
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+int bt_mics_client_write_mute(struct bt_conn *conn, bool mute)
```

Comment:
Is there a reason for this function and the next one being placed in between the discovery functions?

Reply:
Not at all; moved to where it's more relevant.

Reply:
And since I am pestering you with nitpicking, you can consider moving the mics_client_reset as well ;-)

Reply:
I think `mics_client_reset` is placed alright at the moment (static function, right above the other function that calls it).

Reply:
You are right - it is called earlier than I thought.  (I should have checked, not assumed.)